rust/template/differential_datalog/src/program/mod.rs

```diff
@@ -1274,19 +1278,118 @@ impl Program {
         }
     }
 
-    fn xform_collection_ref<'a, 'b, P, T, LC>(
-        col: &Collection<Child<'a, P, T>, DDValue, Weight>,
+    fn xform_collection_ref<'a, S, T, Lookup>(
+        col: &Collection<S, DDValue, Weight>,
+        xform: &XFormCollection,
+        arrangements: &Arrangements<'a, S, T>,
+        lookup_collection: Lookup,
+        is_top_level_scope: bool,
+    ) -> Collection<S, DDValue, Weight>
+    where
+        S: Scope,
+        S::Timestamp: Lattice + Refines<T> + ToTupleTS,
+        T: Lattice + Timestamp,
+        Lookup: Fn(RelId) -> Option<Collection<S, DDValue, Weight>>,
+    {
+        if let XFormCollection::StreamXForm {
+            description,
+            xform,
+            next,
+        } = xform
+        {
+            let xformed = col
+                .scope()
+                .scoped::<AltNeu<S::Timestamp>, _, _>(description, |inner| {
+                    let d_col = with_prof_context(
+                        format!("differentiate stream before: {}", description).as_ref(),
+                        || col.differentiate(inner),
+                    );
+
+                    fn dummy_lookup_collection<S: Scope>(
+                        _: RelId,
+                    ) -> Option<Collection<S, DDValue, Weight>> {
+                        None
+                    }
+
+                    // We must call the streamless variant within the nested scope
+                    // otherwise we force rustc to instantiate an infinitely long type
+                    // since the function calls itself (a potentially infinite number of times),
+                    // each requiring further nesting of the scopes (and their types)
+                    let xformed = Self::streamless_xform_collection::<
+                        Child<S, AltNeu<S::Timestamp>>,
+                        S::Timestamp,
+                        _,
+                    >(
+                        d_col,
+                        &*xform,
+                        &Arrangements {
+                            arrangements: &FnvHashMap::default(),
+                        },
+                        dummy_lookup_collection,
+                        false,
+                    );
+
+                    with_prof_context(
+                        format!("integrate stream after: {}", description).as_ref(),
+                        || xformed.integrate(),
+                    )
+                });
+
+            Self::xform_collection(
+                xformed.clone(),
+                &*next,
+                arrangements,
+                lookup_collection,
+                is_top_level_scope,
+            )
+        } else {
+            Self::streamless_xform_collection_ref(
```

Comment:
This might be wrong.  This assumes that only the first operator in a rule can be a stream xform and will break otherwise.  In practice, this means that if a streaming group_by occurs after some other operators, e.g., streaming join, things will break.  I will try to create a counterexample, but may not have the time until later today.